qa/L0_response_cache/ensemble_cache_test.py

```diff
@@ -0,0 +1,296 @@
+        self.model_directory = os.path.join(os.getcwd(), "models", "ensemble_models")
+        self.ensemble_config_file = os.path.join(
+            self.model_directory, self.ensemble_model, "config.pbtxt"
+        )
+        self.composing_config_file = os.path.join(
+            self.model_directory, self.composing_model, "config.pbtxt"
+        )
+        input0_data = np.ones((1, 16), dtype=np.float32)
+        input1_data = np.ones((1, 16), dtype=np.float32)
+        self.input_tensors = [
+            grpcclient.InferInput(
+                "INPUT0", input0_data.shape, np_to_triton_dtype(input0_data.dtype)
+            ),
+            grpcclient.InferInput(
+                "INPUT1", input1_data.shape, np_to_triton_dtype(input0_data.dtype)
+            ),
+        ]
+        self.input_tensors[0].set_data_from_numpy(input0_data)
+        self.input_tensors[1].set_data_from_numpy(input1_data)
+
+    def _update_config(self, config_file, config_pattern, config_to_add):
+        # Utility function to update config files as per testcase
+        with open(config_file, "r") as f:
+            config_data = f.read()
+            if config_pattern not in config_data:
+                with open(config_file, "w") as f:
+                    config_data += config_to_add
+                    f.write(config_data)
+
+    def _remove_config(self, config_file, config_to_remove):
+        # Utility function to remove extra added config from the config files
+        with open(config_file, "r") as f:
+            config_data = f.read()
+        updated_config_data = re.sub(config_to_remove, "", config_data)
+        with open(config_file, "w") as f:
+            f.write(updated_config_data)
+
+    def _reset_config_files(self):
+        # Utillity function to reset all config files to original
+        self._remove_config(self.ensemble_config_file, RESPONSE_CACHE_CONFIG)
+        self._remove_config(self.composing_config_file, RESPONSE_CACHE_CONFIG)
+
+    def _run_ensemble(self):
+        # Run the ensemble pipeline and validate output
+        output = self.triton_client.infer(
+            model_name=self.ensemble_model, inputs=self.input_tensors
+        )
+        self.assertIsNotNone(
+            output,
+            f"Unexpected error: Inference result is None for model '{self.ensemble_model}'. Expected non-null output.",
+        )
+        output0 = output.as_numpy("OUTPUT0")
+        output1 = output.as_numpy("OUTPUT1")
+        outputs = [output0, output1]
+        return outputs
+
+    def _get_model_statistics(self, model):
+        # Get the stats for the requested model
+        model_stats = self.triton_client.get_inference_statistics(
+            model_name=model, as_json=True
+        )
+        return model_stats["model_stats"][1]["inference_stats"]
+
+    def _run_inference_and_validate(self, model):
+        # Utility function to call run_ensemble for inference and validate expected baseline output and stats
```

Comment:
May I ask to specify the test plan for this sub-routine? Something like this: https://github.com/triton-inference-server/server/blob/20627268b9f7e951e8be7f49f0d28bf76f178961/qa/L0_trace/opentelemetry_unittest.py#L236-L250

This way it would be clearer why we load models in question and what this routine is doing

Reply:
Sure, will add it.